libtransmission/global-ip-cache.cc

```diff
@@ -0,0 +1,375 @@
+{
+    auto lock = std::unique_lock{ is_updating_mutex_[type] };
+    is_updating_cv_[type].wait(
+        lock,
+        [this, type]() { return is_updating_[type] == is_updating_t::NO || is_updating_[type] == is_updating_t::ABORT; });
+    if (is_updating_[type] != is_updating_t::NO)
+    {
+        return false;
+    }
+    is_updating_[type] = is_updating_t::YES;
+    lock.unlock();
+    is_updating_cv_[type].notify_one();
+    return true;
+}
+
+void tr_global_ip_cache::unset_is_updating(tr_address_type type) noexcept
+{
+    TR_ASSERT(is_updating_[type] == is_updating_t::YES);
+    auto lock = std::unique_lock{ is_updating_mutex_[type] };
+    is_updating_[type] = is_updating_t::NO;
+    lock.unlock();
+    is_updating_cv_[type].notify_one();
+}
+
+void tr_global_ip_cache::update_global_addr(tr_address_type type) noexcept
+{
+    TR_ASSERT(has_ip_protocol_[type]);
+    TR_ASSERT(global_source_addr(type));
+    TR_ASSERT(ix_service_[type] < std::size(IPQueryServices));
+
+    if (ix_service_[type] == 0U && !set_is_updating(type))
+    {
+        return;
+    }
+    TR_ASSERT(is_updating_[type] == is_updating_t::YES);
+
+    // Update global address
+    auto options = tr_web::FetchOptions{ IPQueryServices[ix_service_[type]],
+                                         [this, type](tr_web::FetchResponse const& response)
+                                         { this->on_response_ip_query(type, response); },
+                                         nullptr };
+    options.ip_proto = type == TR_AF_INET ? tr_web::FetchOptions::IPProtocol::V4 : tr_web::FetchOptions::IPProtocol::V6;
+    options.sndbuf = 4096;
+    options.rcvbuf = 4096;
+    web_.fetch(std::move(options));
+}
+
+void tr_global_ip_cache::update_source_addr(tr_address_type type) noexcept
+{
+    TR_ASSERT(has_ip_protocol_[type]);
+
+    if (!set_is_updating(type))
+    {
+        return;
+    }
+    TR_ASSERT(is_updating_[type] == is_updating_t::YES);
+
+    auto const protocol = type == TR_AF_INET ? "IPv4"sv : "IPv6"sv;
+
+    auto err = int{};
+    auto const& source_addr = get_global_source_address(bind_addr(type), err);
+    if (source_addr)
+    {
+        set_source_addr(*source_addr);
+        tr_logAddInfo(fmt::format(
+            _("Successfully updated source {protocol} address to {ip}"),
+            fmt::arg("protocol", protocol),
+            fmt::arg("ip", source_addr->display_name())));
+    }
+    else
+    {
+        // Stop the update process since we have no public internet connectivity
+        unset_addr(type);
+        upkeep_timers_[type]->setInterval(RetryUpkeepInterval);
+        tr_logAddDebug(fmt::format(_("Couldn't obtain source {protocol} address"), fmt::arg("protocol", protocol)));
+        if (err == EAFNOSUPPORT)
+        {
+            stop_timer(type); // No point in retrying
+            has_ip_protocol_[type] = false;
+            tr_logAddWarn(fmt::format(_("Your machine does not support {protocol}"), fmt::arg("protocol", protocol)));
```

Comment:
_(nit)_

Should be an `Info` instead of a `Warn`; it's reasonable to have an IPv4-only setup